src/notation/internal/notationparts.cpp

```diff
@@ -706,8 +706,8 @@ void NotationParts::replaceInstrument(const ID& instrumentId, const ID& fromPart
         return;
     }
 
-    part->setInstrument(InstrumentsConverter::convertInstrument(newInstrument), oldInstrumentInfo.fraction);
-    doSetPartName(part, formatPartName(part));
+    score()->undo(new Ms::ChangePart(part, new Ms::Instrument(InstrumentsConverter::convertInstrument(newInstrument)),
+                                     formatPartName(part)));
```

Comment:
for Undo to work properly we can't change the instrument on the part before creating the "ChangePart" command, as it needs access to the original value in order to revert to it when you undo. ChangePart can also change the name, so no need to call doSetPartName explicitly.